xfrd-tcp.c

```diff
@@ -24,6 +24,82 @@
 #include "xfrd.h"
 #include "xfrd-disk.h"
 #include "util.h"
+#ifdef HAVE_TLS_1_3
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+#endif
+
+#ifdef HAVE_TLS_1_3
+static SSL_CTX*
+create_ssl_context()
+{
+	SSL_CTX *ctx;
+	ctx = SSL_CTX_new(TLS_client_method());
+	if (!ctx) {
+		log_msg(LOG_ERR, "xfrd tls: Unable to create SSL ctxt");
+	}
+	else if (SSL_CTX_set_default_verify_paths(ctx) != 1) {
+		SSL_CTX_free(ctx);
+		log_msg(LOG_ERR, "xfrd tls: Unable to set default SSL verify paths");
+	}
+	/* Only trust 1.3 as per the specification */
+	else if (!SSL_CTX_set_min_proto_version(ctx, TLS1_3_VERSION)) {
+		SSL_CTX_free(ctx);
+		log_msg(LOG_ERR, "xfrd tls: Unable to set minimum TLS version 1.3");
+	}
+	return ctx;
+}
+
+static int
+tls_verify_callback(int preverify_ok, X509_STORE_CTX *ctx)
+{
+	int err = X509_STORE_CTX_get_error(ctx);
+	int depth = X509_STORE_CTX_get_error_depth(ctx);
+
+	// report the specific cert error here - will need custom verify code if 
+	// SPKI pins are supported
+	if (!preverify_ok)
+		log_msg(LOG_ERR, "xfrd tls: TLS verify failed - (%d) depth: %d error: %s",
+				err,
+				depth,
+				X509_verify_cert_error_string(err));
+	return preverify_ok;
+}
+
+static int
+setup_ssl(struct xfrd_tcp_pipeline* tp, struct xfrd_tcp_set* tcp_set, 
+          const char* auth_domain_name)
+{
+	if (!tcp_set->ssl_ctx) {
+		log_msg(LOG_ERR, "xfrd tls: No TLS CTX, cannot set up XFR-over-TLS");
+		return 0;
+	}
+	DEBUG(DEBUG_XFRD,1, (LOG_INFO, "xfrd: setting up TLS for tls_auth domain name %s", 
+						 auth_domain_name));
+	tp->ssl = SSL_new((SSL_CTX*)tcp_set->ssl_ctx);
+	if(!tp->ssl) {
+		log_msg(LOG_ERR, "xfrd tls: Unable to create TLS object");
+		return 0;
+	}
+	SSL_set_connect_state(tp->ssl);
+	(void)SSL_set_mode(tp->ssl, SSL_MODE_AUTO_RETRY);
+	if(!SSL_set_fd(tp->ssl, tp->tcp_w->fd)) {
+		log_msg(LOG_ERR, "xfrd tls: Unable to set TLS fd");
+		SSL_free(tp->ssl);
+		return 0;
+	}
```

Comment:
Set the ssl pointer to NULL like in the other code path before error exit from the function.

```suggestion
	if(!SSL_set_fd(tp->ssl, tp->tcp_w->fd)) {
		log_msg(LOG_ERR, "xfrd tls: Unable to set TLS fd");
		SSL_free(tp->ssl);
		tp->ssl = NULL;
		return 0;
	}
```

Reply:
Resolved in current code.